### ISSUE TITLE
[Core] Tx Fee Business Logic - Improve State Visibility

### ISSUE THREAD

**Issue by jessicadaugherty** [pokt-network/pocket#325]
## Objective

Validate and expose fees throughout the blockchain state when a transaction (successful or not) is sent.

## Origin Document

The flow shown in the diagram below, is an example where funds where sent from one account to another and fees were  transmitted to the associated pool from the sender's account balance

![Image](https://user-images.githubusercontent.com/55156619/213946403-c46c7ecc-356c-419c-93f1-0ed503642dfe.png)

We can validate the status of transactions and fees using the [logging module.](https://github.com/pokt-network/pocket/issues/288) 

## Goals

- Verify that state-related changes related to fees and balances are correct when a transaction is sent

## Deliverable

- [ ] Define and add tooling to increase visibility into the state of the database (balances of accounts and fees)
- [ ] Documentation on how LocalNet should be used to manually test an end-to-end flow, validating balances & fees while doing so

## Non-goals / Non-deliverables

- Adding new business logic into utility 
- Adding tx to mempool

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)

## Testing Methodology

- **All tests**: `make test_all`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @jessicadaugherty
**Co-owner**: @Olshansk

**Comment by Olshansk** [comment]
@jessicadaugherty Updated the description.

**Comment by jessicadaugherty** [comment]
Is there a reason these things can't be covered in a separate ticket? 

>  Optional: if implementation is incomplete, a PR that finalizes tx fee management
>  A PR that adds additional unit tests for accounts/pool balances & fees when a tx is successful
>  A PR that adds additional unit tests for accounts/pool balances & fees when a tx is no successful

I think this is adding confusion and a distraction from the goal of sending a "happy" tx before designing the "unhappy" tx path.

**Comment by Olshansk** [comment]
> Is there a reason these things can't be covered in a separate ticket?
> 
> > Optional: if implementation is incomplete, a PR that finalizes tx fee management
> > A PR that adds additional unit tests for accounts/pool balances & fees when a tx is successful
> > A PR that adds additional unit tests for accounts/pool balances & fees when a tx is no successful
> 
> I think this is adding confusion and a distraction from the goal of sending a "happy" tx before designing the "unhappy" tx path.

Make sense. Let's add those deliverables to a separate ticket.

**Comment by jessicadaugherty** [comment]
**Design Considerations**

In V0 today, invalid txs take fees. However, invalid txs without enough POKT to cover fees are still stored on-chain per Tendermint which can result in block bloat.

V0 plans on addressing this issue in https://github.com/pokt-network/pocket-core/issues/1505 

Pending confirmation from @oten91, I believe the intention is to take fees and store invalid txs in memory rather than in the block, and to block invalid txs that do not have enough fees to cover the tx).

**Comment by jessicadaugherty** [comment]
New ticket: Productionize E2E tx in mainline. Comes before this validate ticket.

**Comment by Olshansk** [comment]
> New ticket: Productionize E2E tx in mainline. Comes before this validate ticket.

https://github.com/pokt-network/pocket/issues/364

**Comment by Olshansk** [comment]
Discussed the context above offline last week, so just wanted to capture a tl;dr of the discussion

At a high level, we'll need to create several tickets of work before we can dive into the pocket utility design & implementation):

- [ ]  1. Utility - Improve visibility into utility state (i.e. w/ tooling or documentation)
- [ ]  2. Utility - Document tx fee logic (valid, invalid, included in block or not)
- [ ]  3. Utility - Cleanup & refactor existing code (both the scope and changes are included here)
- [ ]  4. Utility - Define testing edge cases for PoS network (to make sure there are no regressions)
- [ ]  5.Utility - Add testing for E2E scenario (potentially using cucumber / gherkin)

This ticket will focus on (1).

**Comment by Olshansk** [comment]
The WIP for this is in the `issues/325/tx_fee_business_logic_visibility.

Neo4j is being used to visualize the fee pools (e.g. DAO), accounts (e.g. Address) and other actors (e.g. Validators) via a visual interface. 

<img width="1608" alt="Screenshot_2023-02-01_at_9 29 28_PM" src="https://user-images.githubusercontent.com/1892194/217042894-55890fc1-7b57-4a78-866e-0b3e403a2ffd.png">
